Snowflake.Data/Core/Session/SessionPool.cs

```diff
@@ -175,34 +190,41 @@ private SessionOrCreationToken GetIdleSession(string connStr)
                         s_logger.Debug("SessionPool::GetIdleSession - no thread was waiting for a session, an idle session was retrieved from the pool");
-                        return new SessionOrCreationToken(session);
+                        return new SessionOrCreationTokens(session);
                     }
                     s_logger.Debug("SessionPool::GetIdleSession - no thread was waiting for a session, but could not find any idle session available in the pool");
-                    if (IsAllowedToCreateNewSession())
+                    var sessionsCount = AllowedNumberOfNewSessionCreations();
+                    if (sessionsCount > 0)
                     {
-                        // there is no need to wait for a session since we can create a new one
-                        return new SessionOrCreationToken(_sessionCreationTokenCounter.NewToken());
+                        // there is no need to wait for a session since we can create new ones
+                        var sessionCreationTokens = Enumerable.Range(1, sessionsCount)
+                            .Select(_ => _sessionCreationTokenCounter.NewToken())
+                            .ToList();
+                        return new SessionOrCreationTokens(sessionCreationTokens);
                     }
                 }
             }
-            return new SessionOrCreationToken(WaitForSession(connStr));
+            return new SessionOrCreationTokens(WaitForSession(connStr));
         }
 
-        private bool IsAllowedToCreateNewSession()
+        private int AllowedNumberOfNewSessionCreations()
         {
             if (!IsMultiplePoolsVersion())
             {
                 s_logger.Debug($"SessionPool - creating of new sessions is not limited");
-                return true;
+                return 1; // waiting disabled means we are either in old pool or there is no pooling
             }
             var currentSize = GetCurrentPoolSize();
             if (currentSize < _poolConfig.MaxPoolSize)
             {
-                s_logger.Debug($"SessionPool - allowed to create a session, current pool size is {currentSize} out of {_poolConfig.MaxPoolSize}");
-                return true;
+                var maxSessionsToCreate = _poolConfig.MaxPoolSize - currentSize;
+                var sessionsNeeded = Math.Max(_poolConfig.MinPoolSize - currentSize, 1);
+                var sessionsToCreate = Math.Min(sessionsNeeded, maxSessionsToCreate);
+                s_logger.Debug($"SessionPool - allowed to create {sessionsToCreate} sessions, current pool size is {currentSize} out of {_poolConfig.MaxPoolSize}");
+                return sessionsToCreate;
             }
             s_logger.Debug($"SessionPool - not allowed to create a session, current pool size is {currentSize} out of {_poolConfig.MaxPoolSize}");
-            return false;
+            return 0;
         }
 
         private bool IsMultiplePoolsVersion() => _waitingForIdleSessionQueue.IsWaitingEnabled();
```

Comment:
comment for line: 404 - please ensure MinPoolSize in AddSession method when old/expired sessions made session below MinPoolSize

Reply:
done